Makefile

```
@@ -30,24 +30,27 @@ all: test build package
 
 build: build-linux build-windows
 
-build-linux-lifecycle: export GOOS:=linux
-build-linux-lifecycle: OUT_DIR:=$(BUILD_DIR)/$(GOOS)/lifecycle
-build-linux-lifecycle: GOENV:=GOARCH=$(GOARCH) CGO_ENABLED=1
-build-linux-lifecycle: DOCKER_RUN=docker run --workdir=/lifecycle -v $(OUT_DIR):/out -v $(PWD):/lifecycle $(COMPILATION_IMAGE)
-build-linux-lifecycle:
+$(BUILD_DIR)/linux/lifecycle/lifecycle: export GOOS:=linux
+$(BUILD_DIR)/linux/lifecycle/lifecycle: OUT_DIR:=$(BUILD_DIR)/$(GOOS)/lifecycle
+$(BUILD_DIR)/linux/lifecycle/lifecycle: GOENV:=GOARCH=$(GOARCH) CGO_ENABLED=1
+$(BUILD_DIR)/linux/lifecycle/lifecycle: DOCKER_RUN=docker run --workdir=/lifecycle -v $(OUT_DIR):/out -v $(PWD):/lifecycle $(COMPILATION_IMAGE)
+$(BUILD_DIR)/linux/lifecycle/lifecycle:
```

```suggestion
GOFILES := $(wildcard *.go */*.go */*/*.go)
$(BUILD_DIR)/linux/lifecycle/lifecycle: export GOOS:=linux
$(BUILD_DIR)/linux/lifecycle/lifecycle: OUT_DIR:=$(BUILD_DIR)/$(GOOS)/lifecycle
$(BUILD_DIR)/linux/lifecycle/lifecycle: GOENV:=GOARCH=$(GOARCH) CGO_ENABLED=1
$(BUILD_DIR)/linux/lifecycle/lifecycle: DOCKER_RUN=docker run --workdir=/lifecycle -v $(OUT_DIR):/out -v $(PWD):/lifecycle $(COMPILATION_IMAGE)
$(BUILD_DIR)/linux/lifecycle/lifecycle: $(GOFILES)
```
To make these target rebuild when the source files change you need to do something similar to this ^.

There might be a way to make the wildcard work for an arbitrary depth. Alternatively some `go list -json` + `jq` magic could also get you the source file list as well.

---

This could be improved to do things like exclude test files. Also, given that we want a very limited set of imports for launcher we could scope the source file list down for the launcher build targets (or we could just be lazy for now and use the same list).

---

The latest changes use `wilcard` but the above mentioned concerns (arbitrary depth, excluding test files, improving launcher) remain (see discussion in [Slack](https://buildpacks.slack.com/archives/CSFE49VGR/p1594309841165100)). Thanks @jromero for your suggestions, maybe we can yet get `go list` to do our bidding.

---

Thanks to @jromero this is working now with `go list`. I am going to be lazy for now and say that we should punt on optimizing the launcher.